### PR TITLE
Added default locale before login

### LINKED ISSUE #8641: Translation of App login screen not working
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

When already logged in and opening directus app page, the login screen is not translated 
 

### To Reproduce

Open directus App page, the login screen is not translated ("Sign in", "Sign out", "Continue", "authenticated" should be in french) : 

step1

If I click on "continue" and then click on the back button of the browser, the login screen is correctly translated (in french in my example) :

step2

### What version of Directus are you using?

v9.0.0-rc96

### What version of Node.js are you using?

12.18.1

### What database are you using?

MySQL

### What browser are you using?

Firefox

### What operating system are you using?

CentOS

### How are you deploying Directus?

locally

### PR DESCRIPTION
**Pull request by christianrr**
Closes #7062

### CODE REVIEW

**Review by nickrum — COMMENTED** [review]
As mentioned in the discussion, I think it would be more convenient to fall back to `project_language` when the users' language isn't set, instead of overwriting it.

**Comment by christianrr** [comment]
> As mentioned in the discussion, I think it would be more convenient to fall back to `project_language` when the users' language isn't set, instead of overwriting it.

How can a user's language not be set? The users' language field is mandatory and whenever a new user is created in the Admin app we start off with the project_language as default - until a language is chosen by the user!? Sorry, I don't exactly understand what you mean...

**Comment by nickrum** [comment]
Sorry, I should have been more specific. As mentioned in the linked discussion, I think we should make the user's language nullable and make `null` the default value. If the user's language is set to `null`, the language of Directus should then fall back to the project language. This way the default language of all users can be changed after they have been created, unless they set their lanugage explicitly.

**Comment by benhaynes** [comment]
I like that, good idea @nickrum! We'll want to update the interface so `NULL` shows "Project Default".

**Comment by christianrr** [comment]
kk, I'll look into it next week!

**Review by nickrum — CHANGES_REQUESTED** [review]

**Comment by neo7-studio-web** [comment]
Any chance to have it resolved in the v9 stable version ?

File: app/src/lang/available-languages.yaml
Comment (christianrr) [reply app/src/lang/available-languages.yaml]: Agreed, added all requested changes in recent commits

**Comment by christianrr** [comment]
@nickrum Is there anything else todo for me? I committed the requested changes 16 days ago... Are other tasks higher prioritized at the moment, or do I miss something?

**Comment by benhaynes** [comment]
> @nickrum Is there anything else todo for me? I committed the requested changes 16 days ago... Are other tasks higher prioritized at the moment, or do I miss something?

All PRs require multiple members of the core team to review before merging. However, the core team has been extremely busy with the release to GA (announced yesterday), the release of our new Cloud, and stability/fixes. This will get reviewed/merged, but is just not at the top of the list.

The good news is that after months of work on the tasks I mentioned, we're transitioning the team to issues and PRs full-time now... so the long delay is over and we're actively working through the list.

**Comment by christianrr** [comment]
> > @nickrum Is there anything else todo for me? I committed the requested changes 16 days ago... Are other tasks higher prioritized at the moment, or do I miss something?
> 
> All PRs require multiple members of the core team to review before merging. However, the core team has been extremely busy with the release to GA (announced yesterday), the release of our new Cloud, and stability/fixes. This will get reviewed/merged, but is just not at the top of the list.
> 
> The good news is that after months of work on the tasks I mentioned, we're transitioning the team to issues and PRs full-time now... so the long delay is over and we're actively working through the list.

Ok, thanks for clarifying... You all do a great job, many thanks!

**Comment by neo7-studio-web** [comment]
Any news about this one ?

**Comment by rijkvanzanten** [comment]
> Any news about this one ?

You'd hear it here first! No activity on the PR is no updates on the PR 👍🏻

**Comment by loongmxbt** [comment]
Same issue, where to set project default language. (login page translation)

**Comment by christianrr** [comment]
hmm, the database migration fails to set null as default for the language field in directus_users for Oracle:
`table.string('language').nullable().alter();`

I'm not an expert on Knex migrations, how would I implement that correctly?

**Comment by jaycammarano** [comment]
> hmm, the database migration fails to set null as default for the language field in directus_users for Oracle: `table.string('language').nullable().alter();`
> 
> I'm not an expert on Knex migrations, how would I implement that correctly?

I was going to ask @aidenfoxx

**Comment by aidenfoxx** [comment]
@christianrr I see that `directus_users.language` is seeded as nullable. Assuming that isn't changed by any other migration then your migration is trying to set nullable to nullable, which Oracle rejects? If that's the case, I think you just need to remove that from the new migration?

**Comment by aidenfoxx** [comment]
I don't know what the solution is then. I assumed `.defaultTo(null)` would work for Oracle. What was the error?

**Comment by christianrr** [comment]
@aidenfoxx Reverted now to .defaultTo(null) - error says:
`error: alter table "directus_users" alter column "language" type varchar(255) using ("language"::varchar(255)) - ALTER COLUMN TYPE from varchar to varchar is only supported experimentally`

**Comment by aidenfoxx** [comment]
Ah yeah... that error! Knex always generates the alter queries with the type even if you just want to change the default. And I don't think `setNullable` will work since it'll retain the existing default.

This is probably something to bring up in the https://github.com/knex/knex repo since there should be some way to remove a default!

**Comment by christianrr** [comment]
Ok, thx for the info. I read in some forum where somebody used a raw query for altering the default value? But that wouldn't work for all database types because of different syntax, right?

Any other suggestions?

https://stackoverflow.com/questions/41959299/my-sql-alter-table-using-knex-js

**Comment by aidenfoxx** [comment]
You could always add a Oracle specific query as we've done in the past: https://github.com/directus/directus/commit/e8e295ef6a208fcaa391ee1ce80fc67c731ebf63#diff-3bbba2b75650f2c9a7e5f1d3aa5d7bbc5b1f4e371998e122706313b93ba36edd

Or alternatively try the new `changeToTypeByCopy` helper function.

**Comment by benhaynes** [comment]
> This is probably something to bring up in the https://github.com/knex/knex repo since there should be some way to remove a default!

Luckily we have a resident maintainor!! @rijkvanzanten 🙂

**Comment by aidenfoxx** [comment]
@christianrr @rijkvanzanten I just had a look into this and I believe you mistook a cockroachdb error for an Oracle error. Oracle should support this correctly (though I did find and fix https://github.com/directus/directus/pull/11359 in the process 😬).

**Comment by christianrr** [comment]
@rijkvanzanten I tried to fix the failing migration test, however there were some recent conflicts. I have tried to merge the most recent main branch into my feature branch and resolve the conflicts, however I have the feeling that I screwed up something. My last (merge) commits are not appearing any more in the PR? Or does it now require an action from you? Sorry, I'm a bit lost ;)

**Comment by christianrr** [comment]
@rijkvanzanten @nickrum Everything seems to be OK now, any change to get this merged any time soon?

File: api/src/database/migrations/20211004A-add-project-language.ts
Comment (rijkvanzanten) [inline api/src/database/migrations/20211004A-add-project-language.ts:13]: Lets use the helpers for this, like:

https://github.com/directus/directus/blob/7d391ed4fad599240d0518fc2ab2c4bc9f20c7bf/api/src/database/migrations/20220303A-remove-default-project-color.ts#L7-L11

File: api/src/database/migrations/20211004A-add-project-language.ts
Comment (rijkvanzanten) [inline api/src/database/migrations/20211004A-add-project-language.ts:8]: Lets call the field `default_language` instead of `project_language` to avoid confusion with the languages configured by the individual users

File: api/src/database/seeds/03-users.yaml
Comment (rijkvanzanten) [inline api/src/database/seeds/03-users.yaml:37]: We shouldn't change seeds (ever). Lets keep this on `en-US`

File: app/src/app.vue
Comment (rijkvanzanten) [inline app/src/app.vue:105]: IIRC the app store hydrates after user login. The serverStore is hydrated _before_ the user logs in. Ideally, we set the correct language after the serverStore is hydrated, so we can show the login page in the correct language as well

File: app/src/app.vue
Comment (rijkvanzanten) [inline app/src/app.vue:115]: This also happens in the hydration for the usersstore if i'm not mistaken. This might be double

File: app/src/interfaces/_system/system-language/system-language.vue
Comment (rijkvanzanten) [inline app/src/interfaces/_system/system-language/system-language.vue:29]: Why would we want to exclude the default language from the list of options? 🤔

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Thanks @christianrr! Couple quick questions:

File: app/src/interfaces/_system/system-language/system-language.vue
Comment (christianrr) [reply app/src/interfaces/_system/system-language/system-language.vue:29]: This is because of the settings page: `system-language` is used as interface for the Default/Project Language field. It does not make sense to be able to choose the "Default Language" option in the dropdown as value for the Default Language field.

Comment (rijkvanzanten) [reply app/src/interfaces/_system/system-language/system-language.vue:29]: Ohh gotcha, yeah that makes sense 👍🏻 I thought the flag would control whether or not the _language_ would show up (so like it would exclude en-US from the list f.e.). Didn't realize there was a dedicated "Project Default" option in that interface 👍🏻

File: app/src/app.vue
Comment (christianrr) [reply app/src/app.vue:115]: This is needed when the user's language is reset to Default Language on the user's detail page. However, you are right - it would be better to handle it there. I will do so.

File: app/src/app.vue
Comment (christianrr) [reply app/src/app.vue:105]: Yeah, you are right. I did quite some testing however and it needs both watches: To watch the server store hydration for setting the language of the login screen and in a second stage the app hydration to set the language when the user's language is known.

File: api/src/database/seeds/03-users.yaml
Comment (christianrr) [reply api/src/database/seeds/03-users.yaml:37]: Ok, I'll revert this.

File: app/src/app.vue
Comment (rijkvanzanten) [reply app/src/app.vue:105]: > Yeah, you are right. I did quite some testing however and it needs both watches: To watch the server store hydration for setting the language of the login screen and in a second stage the app hydration to set the language when the user's language is known.

Yeah. Just curious how it ties in with the existing `setLanguage` calls, and making sure we're not doing the same thing redundantly in multiple places:

https://github.com/directus/directus/blob/ef18cdc14470aea5d6e35d0f4c1ee91aba71d132/app/src/hydrate.ts#L78

https://github.com/directus/directus/blob/ef18cdc14470aea5d6e35d0f4c1ee91aba71d132/app/src/modules/users/routes/item.vue#L438-L449

Comment (christianrr) [reply app/src/app.vue]: You are right, for app hydration setLanguage is called twice.

Would you prefer to have the initial setLanguage calls in hydrate.ts and server.ts (server store) or to handle it via watchers in app.vue?

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**
